### PR TITLE
Feature Statemachine

### PR DESCRIPTION
**Pull request by Adesii**
This implements a basic Turn StateMachine into the Bowling Gamemode.
The StateMachine can be used as a starting point for GameSpecifc StateMachines and doesn't have to be just for Turns.

### CODE REVIEW

File: code/Base/StateSystem/StateMachine.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+			CurrentState.Parent = this;
+			States.Add( name, CurrentState );
+		}
+	}
+
+	protected void PreSpawnEntities( string StartState )
+	{
+		if ( Host.IsClient )
+			return;
+		var FirstPredictionState = TypeLibrary.GetAttribute<PredictStatesAttribute>( TypeLibrary.GetTypeByName( StartState ) );
```
Comment (DevulTj) [inline code/Base/StateSystem/StateMachine.cs:63]: firstPredictionState

File: code/Base/StateSystem/StateMachine.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+		if ( !States.ContainsKey( StartState ) )
+			CacheState( StartState );
```
Comment (DevulTj) [inline code/Base/StateSystem/StateMachine.cs:65]: can we add a new line under this

File: code/Base/StateSystem/StateMachine.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		foreach ( var item in FirstPredictionState.PredictedStates )
+		{
+			CacheState( item );
+		}
+	}
+
+	private void CacheState( string name )
+	{
+		if ( States.ContainsKey( name ) )
+			return;
```
Comment (DevulTj) [inline code/Base/StateSystem/StateMachine.cs:75]: can we add a new line under this

File: code/Base/StateSystem/TurnSystem/TurnStateMachine.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
+using Sports.StateSystem;
+
+namespace Sports;
+
+public partial class TurnStateMachine : StateMachine
+{
+
+	[Net] public List<Client> TurnOrder { get; set; } = new();
+	[Net, Predicted] public bool TurnFinished { get; set; } = false;
```
Comment (DevulTj) [inline code/Base/StateSystem/TurnSystem/TurnStateMachine.cs:9]: can we have attributes on a separate line from properties, please

File: code/Base/StateSystem/TurnSystem/TurnStateMachine.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+				return null;
+			if ( TurnIndex >= TurnOrder.Count )
+			{
+				TurnIndex = 0;
+			}
+			return TurnOrder[TurnIndex];
+		}
+	}
+
+	[Net, Predicted] public int TurnIndex { get; set; }
```
Comment (DevulTj) [inline code/Base/StateSystem/TurnSystem/TurnStateMachine.cs:24]: same here (can we have attributes on a separate line from properties, please)

File: code/Base/StateSystem/TurnSystem/states/WaitState.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
+namespace Sports.StateSystem;
+
+[Library]
+[PredictStates( nameof( TurnState ) )]
+public partial class WaitState : BaseState<TurnStateMachine>
+{
+	[Net, Predicted] public TimeSince CreationTime { get; set; }
```
Comment (DevulTj) [inline code/Base/StateSystem/TurnSystem/states/WaitState.cs:7]: can we have attributes on a separate line from properties, please

File: code/Base/StateSystem/PredictionStateAttribute.cs [hunk 7]
```diff
@@ -0,0 +1,11 @@
+namespace Sports.StateSystem;
+
+[AttributeUsage( AttributeTargets.Class )]
+public class PredictStatesAttribute : Attribute
```
Comment (DevulTj) [inline code/Base/StateSystem/PredictionStateAttribute.cs]: I wonder if we can name this something better. It's not super self explanatory of what this really does unless you know the system

File: code/Base/StateSystem/BaseState.cs [hunk 8]
```diff
@@ -0,0 +1,22 @@
+namespace Sports.StateSystem;
+
+
+public partial class BaseState : Entity
+{
+	[Net] public StateMachine StateMachine { get; set; }
```
Comment (DevulTj) [inline code/Base/StateSystem/BaseState.cs:6]: can we have attributes on a separate line from properties, please

File: code/Base/StateSystem/StateMachine.cs [hunk 9]
```diff
@@ -0,0 +1,90 @@
+namespace Sports.StateSystem;
+
+public partial class StateMachine : Entity
+{
+
+	[Net] public Dictionary<string, BaseState> States { get; set; }
+	[Net] public BaseGamemode Gamemode { get; set; }
+	[Net, Predicted] private BaseState currentState { get; set; }
```
Comment (DevulTj) [inline code/Base/StateSystem/StateMachine.cs:8]: can we have attributes on a separate line from properties, please

File: code/Base/StateSystem/TurnSystem/TurnStateMachine.cs [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+			TurnOrder.Add( item );
+		}
+		TurnIndex = 0;
+		SetState( nameof( TurnState ) );
+	}
+	public virtual void EndGame()
+	{
+		SetState( nameof( LobbyState ) );
+	}
+	public override void Simulate( Client cl )
```
Comment (DevulTj) [inline code/Base/StateSystem/TurnSystem/TurnStateMachine.cs:45]: new line under all method scope endings

File: code/Base/StateSystem/TurnSystem/states/WaitState.cs [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+	public override void CheckSwitchState()
+	{
+		if ( CreationTime > 5 )
+		{
+			StateMachine.TurnIndex += 1 % StateMachine.TurnOrder.Count;
+			StateMachine.TurnFinished = false;
+			StateMachine.SetState( nameof( TurnState ) );
+		}
+	}
+	public override void OnTick()
```
Comment (DevulTj) [inline code/Base/StateSystem/TurnSystem/states/WaitState.cs:22]: new line under all method scope endings

File: code/Gamemodes/Bowling/Bowling.cs [hunk 12]
```diff
@@ -1,15 +1,84 @@
...
 [HammerEntity]
 [Library( "sports_bowling_gamemode" )]
 [EditorModel( "models/Bowling/bpin.vmdl" )]
 [Title( "Bowling Gamemode" )]
 [Category( "Gamemodes" )]
 [Sphere( 128f, 0, 125, 255 )]
-public partial class Bowling : BaseGamemode
+public partial class Bowling : BaseGamemode, IStateMachine<TurnStateMachine>
 {
+	[Net] public TurnStateMachine StateMachine { get; set; }
```
Comment (DevulTj) [inline code/Gamemodes/Bowling/Bowling.cs:13]: new line for attributes

File: code/Gamemodes/Bowling/Bowling.cs [hunk 13]
```diff
@@ -1,15 +1,84 @@
...
 	public override BasePlayer CreatePawn()
 	{
 		return new BowlingPlayer();
 	}
+	public override void Spawn()
+	{
+		base.Spawn();
+		StateMachine = new()
+		{
+			Gamemode = this,
```
Comment (DevulTj) [inline code/Gamemodes/Bowling/Bowling.cs:24]: Maybe State Machine ctor should just pass in the gamemode instead of an object initializer

**Review by DevulTj — CHANGES_REQUESTED** [review]

File: code/Base/StateSystem/PredictionStateAttribute.cs (hunk 7 above)
Comment (Adesii) [reply code/Base/StateSystem/PredictionStateAttribute.cs]: PredictNextStates? CacheStates? CachePredictStates?

i had trouble finding names too...

File: code/Gamemodes/Bowling/Bowling.cs (hunk 13 above)
Comment (Adesii) [reply code/Gamemodes/Bowling/Bowling.cs:24]: does that work with entities? im always scared to use the constructor for entities.

Comment (DevulTj) [reply code/Gamemodes/Bowling/Bowling.cs:24]: should be okay as long as you have an empty ctor as well

Comment (DevulTj) [reply code/Gamemodes/Bowling/Bowling.cs:24]: by empty I mean no args

File: code/Base/StateSystem/PredictionStateAttribute.cs (hunk 7 above)
Comment (DevulTj) [reply code/Base/StateSystem/PredictionStateAttribute.cs]: it's fine for now it's as simple as just renaming it down the line

**Review by DevulTj — APPROVED** [review]

**Pull request merged**